### ISSUE TITLE
New Feature: Focus Mode

### ISSUE THREAD

**Issue by mMoliere** [majlis-erc/manuForma#50]
@wsalesky Ronny and Gregor want to make the tool accessible to users who have no training in TEI whatsoever. One of their critiques is that the interface is too cluttered and that users do not know what is required of them. 

My idea to solve this issue is to modify the display. Users can choose between two viewing modes: 1) the one we have now ("Classic Mode") and a wizard tool. The wizard breaks up the subforms into discrete steps that our users walk through. On every step the users get a clearly defined task and the relevant section of the subform is displayed. Once the step is completed, they move on to the next step.

To be clear, I am not suggesting that we subdivide the forms, but rather to modify the display.

I would provide the division into steps.

But my question to you is, **do think this can be done on a technical level?**

**Comment by wsalesky** [comment]
Something like a carousel for each step? 
https://getbootstrap.com/docs/4.2/components/carousel/

**Comment by mMoliere** [comment]
@wsalesky Yes, a carousel that you can navigate with controls and indicators.

**Comment by wsalesky** [comment]
@mMoliere Okay. That part is the easy part. The hard part will be how the forms are broken up into steps. Let me know what you were thinking. You can even send me some very general mockups if you like.

**Comment by mMoliere** [comment]
@wsalesky Great, then I will get to work and try to break the forms into steps. I will get back to you once, I have something.

**Comment by mMoliere** [comment]
@wsalesky Ronny has modified the idea: He envisions that the subforms be broken down into collapsible headlines. Users click onto the headline, they expand, and they can work on them. We created a short mockup of how this would like in one subform. 

Ronny would also reorder items in the forms to make them more intuitive to the way scholars in Jewish studies are trained. I keep telling him that TEI constrains the order and structure of elements. Is there a way that we can change the order that fields are displayed without impeding the TEI structure itself? (One example in this mockup is the field `repository` which the TEI situates after `institution` and Ronny wants to display after `idno`.

(Collapsibles will also be an important feature in the Srophé app for the manuscripts.)

[manuForma_collapsibles.pdf](https://github.com/majlis-erc/manuForma/files/10268857/manuForma_collapsibles.pdf)

**Comment by wsalesky** [comment]
@mMoliere Collapsable sections are doable, at what level? Just the top level? 

Reordering the elements is not something we can do, unless it is changed in the underlying TEI. This is both a strength and limitation of the manuForma. In order to reorder the elements we would have to create bespoke forms for each TEI type, which would take much longer, be slower and less flexible in terms of data types. The form generator uses some generic language to be as flexible as possible and as such can generate any TEI form you like, as long as it follow the underlying TEI structure.

**Comment by mMoliere** [comment]
@wsalesky OK, reordering is off the table.

These sections have yet to be determined. It is not likely that they would follow one rule. The collapsable sections would be at levels that group fields into a convenient size for the users. 

Another complication: Because the sections would often not align with the top level Ronny and Gregor worry that the dropdown mechanism to create new fields would be distracting. They ask if it is possible that the dropdowns only create new fields for the respective collapsable section.

**Comment by wsalesky** [comment]
@mMoliere Why don't we come up with a test section that I can work out the logic on, and see how feasible this will be.

**Comment by mMoliere** [comment]
@wsalesky I have prepare a mockup for the simplest section: msIdentifier. It displays both collapsable sections opened. Ask me if there is something unclear about the picture.
[manuForma_collapsibles-long.pdf](https://github.com/majlis-erc/manuForma/files/10281502/manuForma_collapsibles-long.pdf)

**Comment by wsalesky** [comment]
@mMoliere Just an update, I am working on this and have a version mostly working. I am having trouble with the up/down arrow functionality, but everything else works.

**Comment by mMoliere** [comment]
@wsalesky That sounds amazing! You are the best! Keep me updated.

**Comment by wsalesky** [comment]
@mMoliere I have come to conclusion that this can not be done if you want to retain the up/down arrows. I think it is not very effective without them, because when you add a new element you want to be able to move it around. Thoughts?

**Comment by wsalesky** [comment]
@mMoliere great news! I asked the XSLTForms list for help on this one and I think we have a working solution. I will mock something up early next week and get it to you for testing.

**Comment by mMoliere** [comment]
@wsalesky This is excellent news! I am looking forward to see it.

**Comment by mMoliere** [comment]
@wsalesky Can you give us an update how the new feature is coming along?

**Comment by wsalesky** [comment]
@mMoliere I'm so sorry, I mentioned you on the pull request, but forgot to mention you here, or send an email. 
There is a test case available with this pull request: https://github.com/majlis-erc/manuForma/pull/61

If webhooks are working you should be able to access the form on your server at: 
manuForma/form.xq?form=forms/msSimple/index.xhtml

This only include the identification subform as an example. 
You can follow the pattern in this config.xml file to build other simple forms, or to add elements to the current form: 
https://github.com/majlis-erc/manuForma/blob/main/forms/formGenerator/config-msSimple.xml

Let me know how it looks. 
-Winona

**Comment by wsalesky** [comment]
@mMoliere The code is in the XSLT already, you can generate either type of form, so a simple form will be generated if the elementGroup is found in the config file, otherwise it will be the standard form. If you want to play around with it yourself that is fine, if you want me to generate some additional tests that is fine also, I will just need you to write the config file, so I know how the elements are grouped. 

One note, this does not do nesting, so you can have one top level grouping for each subform.

**Comment by mMoliere** [comment]
@wsalesky To clarify this point: Do you mean the collapsables do not nest inside each other or that there can be no child elements which have their own child elements?

**Comment by wsalesky** [comment]
@mMoliere the collapsable sections can not have child collapsable sections (no nested groups), but the elements within each collapsable will have all the correct child elements as defined in the schema. Hope that makes sense.

**Comment by mMoliere** [comment]
@wsalesky It is looking very promising. The Contents form was coming along nicely. Now I have stumbled a bit. The displays the following message: 

![Bildschirmfoto 2023-02-08 um 17 02 01](https://user-images.githubusercontent.com/73166992/217583615-680f8024-0ed4-4df4-8e63-dad40a1245f5.jpg)

And this is the code:

```
<subform xpath="/tei:TEI/descendant::tei:msDesc/tei:msContents" formName="msContents" formLabel="Content Description" repeat="no">
            <localSchema src="../schemas/mss-msitem.xml"/>
            <globalSchema src="../schemas/mss-msitem.xml"/>
            <xmlTemplate src="../templates/full-mss-template.xml"/>
            <formDesc>Describe the textual contents in detail or in general terms.</formDesc>
            <elementGroups>
                <group groupLabel="Summary of the contents" groupNo="1">
                    <element xpath="tei:summary"/>
                </group>
                <group groupLabel="Basic Text Information (Subject Field, Folios, Author, Title, Language)" groupNo="2">
                    <element xpath="tei:msItem"/>
                    <element xpath="tei:locus"/>
                    <element xpath="tei:title"/>
                    <element xpath="tei:author"/>
                    <element xpath="tei:persName"/>
                    <element xpath="tei:textLang"/>
                </group>
                <group groupLabel="Detailed Text Information (Text Division, Incipit, Explicit, Final Rubric, Colophon, Notes)" groupNo="3">
                    <element xpath="tei:rubric"/>
                    <element xpath="tei:incipit"/>
                    <element xpath="tei:explicit"/>
                    <element xpath="tei:colophon"/>
                    <element xpath="tei:locus"/>
                    <element xpath="tei:q"/>
                    <element xpath="tei:note"/>
                </group>
            </elementGroups>
        </subform>
```

Did I do something illicit to the collapsibles?

**Comment by wsalesky** [comment]
Looks like the XSLT is outputting a duplicate ID. I will get you an update ASAP.

**Comment by wsalesky** [comment]
@mMoliere Checkout the latest version of the XSLT, it should fix the issue with the duplicate IDs. Let me know if you run into any other issues.

**Comment by mMoliere** [comment]
@wsalesky Thank for the quick fix. It did work for a moment. Then, I noticed that there were no elements in the third collapsible and I fiddled with the config. As a result the error is back. Next, I rolled back all my changes to the config, but despite this the error persists.

**Comment by wsalesky** [comment]
@mMoliere Interesting, can you paste the code from your config that was failing into the issue? I will do some more testing.

**Comment by mMoliere** [comment]
```
<subform xpath="/tei:TEI/descendant::tei:msDesc/tei:msIdentifier" formName="msIdentifier" formLabel="Identification" repeat="no">
            <localSchema src="../schemas/mss-msidentifier.xml"/>
            <globalSchema src="../schemas/mss-msidentifier.xml"/>
            <xmlTemplate src="../templates/full-mss-template.xml"/>
            <formDesc>Add shelfmark and information on the holding institution and alternate references to existing cataloging data.</formDesc>
            <controlledValuesElements>
                <element ident="country">
                    <valList type="closed">
                        <valItem ident="Russia"/>
                        <valItem ident="USA"/>
                        <valItem ident="Great Britain"/>
                        <valItem ident="Israel"/>
                        <valItem ident="Hungary"/>
                        <valItem ident="France"/>
                    </valList>
                </element>
                <element ident="institution">
                    <valList type="closed">
                        <valItem ident="National Library of Russia"/>
                        <valItem ident="British Library"/>
                        <valItem ident="Ben Zvi Institute"/>
                        <valItem ident="Library of the Hungarian Academy of Sciences"/>
                        <valItem ident="Cambridge University Library"/>
                        <valItem ident="National Library of France"/>
                        <valItem ident="Alliance Israélite Universelle"/>
                        <valItem ident="Jewish Theological Seminary"/>
                    </valList>
                </element>
                <element ident="settlement">
                    <valList type="closed">
                        <valItem ident="St. Petersburg"/>
                        <valItem ident="London"/>
                        <valItem ident="Jerusalem"/>
                        <valItem ident="Budapest"/>
                        <valItem ident="Cambridge"/>
                        <valItem ident="Paris"/>
                        <valItem ident="New York"/>
                    </valList>
                </element>
                <element ident="collection">
                    <valList type="closed">
                        <valItem ident="Kaufmann"/>
                        <valItem ident="Or."/>
                        <valItem ident="T-S Ar."/>
                    </valList>
                </element>
            </controlledValuesElements>
            <elementGroups>
                <group groupLabel="Identification via location, institution, collection, shelfmark" groupNo="1">
                    <element xpath="tei:country"/>
                    <element xpath="tei:settlement"/>
                    <element xpath="tei:idno"/>
                    <element xpath="tei:institution"/>
                    <element xpath="tei:collection"/>
                </group>
                <group groupLabel="Identification via old shelfmark and name" groupNo="2">
                    <element xpath="tei:altIdentifier"/>
                    <element xpath="tei:msName"/>
                </group>
            </elementGroups>
        </subform>
        <subform xpath="/tei:TEI/descendant::tei:physDesc/tei:objectDesc" formName="objectDesc" formLabel="Codicological Definition" repeat="no">
            <localSchema src="../schemas/mss-definition.xml"/>
            <globalSchema src="../schemas/mss-definition.xml"/>
            <xmlTemplate src="../templates/full-mss-template.xml"/>
            <formDesc>Define the Type of Manuscript.</formDesc>
        </subform>
        <subform xpath="/tei:TEI/descendant::tei:msDesc/tei:msContents" formName="msContents" formLabel="Content Description" repeat="no">
            <localSchema src="../schemas/mss-msitem.xml"/>
            <globalSchema src="../schemas/mss-msitem.xml"/>
            <xmlTemplate src="../templates/full-mss-template.xml"/>
            <formDesc>Describe the textual contents in detail or in general terms.</formDesc>
            <elementGroups>
                <group groupLabel="Summary of the contents" groupNo="1">
                    <element xpath="tei:summary"/>
                </group>
                <group groupLabel="Basic Text Information (Subject Field, Folios, Author, Title, Language)" groupNo="2">
                    <element xpath="tei:msItem"/>
                    <element xpath="tei:locus"/>
                    <element xpath="tei:title"/>
                    <element xpath="tei:author"/>
                    <element xpath="tei:persName"/>
                    <element xpath="tei:textLang"/>
                </group>
                <group groupLabel="Detailed Text Information (Text Division, Incipit, Explicit, Final Rubric, Colophon, Notes)" groupNo="3">
                    <element xpath="tei:rubric"/>
                    <element xpath="tei:incipit"/>
                    <element xpath="tei:explicit"/>
                    <element xpath="tei:colophon"/>
                    <element xpath="tei:locus"/>
                    <element xpath="tei:q"/>
                    <element xpath="tei:note"/>
                </group>
            </elementGroups>
        </subform>
```

**Comment by wsalesky** [comment]
@mMoliere This runs for me. I am going to add a revision # to the XSLT so it is easier to make sure that you have the most up to date version. Most current version is 1.01, line 17 in the xslt.

**Comment by mMoliere** [comment]
@wsalesky Indeed, the error is gone! Thank you!

The next issue is that the elements do not appear inside the correct collapsibles. The elements in the red frame are supposed to be displayed in the third collapsible. 
 
![Screenshot 2023-02-09 at 14-24-45 Edit MS TEI Records](https://user-images.githubusercontent.com/73166992/217827171-096719fd-41e1-4cf2-90d8-7c5366cdc6a1.jpg)

**Comment by mMoliere** [comment]
@wsalesky Adding the path was a good idea. But it didn't work. manuForma gave me another error message. Do you think there is a chance that you can add that extra flexibility?

**Comment by wsalesky** [comment]
@mMoliere I can give it a try, but this will create a problem with repeats. You could only have 1 msItem, I suspect duplicates will cause all sorts of issues, since you can not group them together logically. I will experiment and see what I can come up with.

**Comment by wsalesky** [comment]
@mMoliere i may be able to make it work, as long as all the items in the group are children of the same element. 
So this would be okay: 
```
<elementGroups>
                <group groupLabel="Summary of the contents" groupNo="1">
                    <element xpath="tei:summary"/>
                </group>
                <group groupLabel="Basic Text Information (Subject Field, Folios, Author, Title, Language)" groupNo="2" xpath="tei:msItem">
                    <!--<element xpath="tei:msItem"/>-->
                    <element xpath="tei:locus"/>
                    <element xpath="tei:title"/>
                    <element xpath="tei:author"/>
                    <element xpath="tei:persName"/>
                    <element xpath="tei:textLang"/>
                </group>
                <group groupLabel="Detailed Text Information (Text Division, Incipit, Explicit, Final Rubric, Colophon, Notes)" groupNo="3" xpath="tei:msItem">
                    <element xpath="tei:rubric"/>
                    <element xpath="tei:incipit"/>
                    <element xpath="tei:explicit"/>
                    <element xpath="tei:colophon"/>
                    <element xpath="tei:locus"/>
                    <element xpath="tei:q"/>
                    <element xpath="tei:note"/>
                </group>
            </elementGroups>
```

But this would not: 
```

<group groupLabel="Basic Text Information (Subject Field, Folios, Author, Title, Language)" groupNo="2">
                    <element xpath="tei:msItem"/>
                    <element xpath="tei:msItem/tei:locus"/>
                    <element xpath="tei:msItem/tei:title"/>
                    <element xpath="tei:msItem/tei:author"/>
                    <element xpath="tei:msItem/tei:persName"/>
                    <element xpath="tei:msItem/tei:textLang"/>
                </group>
```

Also, this makes it so that the msItem is not repeatable. What do you think?

**Comment by mMoliere** [comment]
@wsalesky I am afraid this will not work for us. We need msItem as a repeatable element. But thank you for exploring this option. 

Another thing: We would also like the start page to display collapsibles. Coul you implement it from the following mockup?
![image](https://user-images.githubusercontent.com/73166992/218569975-f939f6db-efb4-4511-b9c6-64c6414c5da5.png)

**Comment by wsalesky** [comment]
@mMoliere Okay, that makes sense. In that case I will leave the forms as is. I will work on updates to the start page and have something for you in a day or two.

**Comment by wsalesky** [comment]
@mMoliere what do you want under each tab? currently the start page offers a starting a new record from a template, finding an existing record from the database, or uploading a record. We also have the 'project selector' on the start page. If you can give me a mock up of the two tabs expanded, so I see what you want in each that would be great.

**Comment by mMoliere** [comment]
@wsalesky: 
![manuForma_collapsibles-startpage](https://user-images.githubusercontent.com/73166992/219012023-b799beb7-a050-42ad-b438-70bfed5695b0.jpg)

**Comment by mMoliere** [comment]
@wsalesky The dropdown menu that lets us add new elements is using the TEI nomenclature. Could you restore our custom terms?

**Comment by wsalesky** [comment]
@mMoliere I will take a look.

**Comment by mMoliere** [comment]
@wsalesky When I try to run the XSLT over the config, OxygenXML freezes. I thought that the new subform I added ("reproductions") could be the culprit. But the config file freezes just the same when reproductions is excluded. Can you take a look?

**Comment by wsalesky** [comment]
@mMoliere 
I checked the file, one subform at a time. They all run except the decoDesc subform. I checked the schema and it is missing an elementSpec for decoDesc, once you add that it should run fine. It is really important to have a complete schema for the forms to be built on. That is the first place to check if something goes wrong.

Try the new version and let me know how it runs now.

See: https://github.com/majlis-erc/manuForma/pull/74

**Comment by wsalesky** [comment]
@mMoliere Check out the latest xslt for an update to the main page layout.

**Comment by mMoliere** [comment]
@wsalesky Thank you for the help with the decoDesc schema! It works like a charm now.

I can see no change to the main page. (Should I look out for a new version number?)

In the subform "Codicological Definition" I want to display objectDesc/@form. But I get nothing but a few up-and-down-arrows with no text. What can we do? 
![Screenshot 2023-02-15 at 21-32-30 Edit MS TEI Records](https://user-images.githubusercontent.com/73166992/219147855-64b89da7-abd8-444c-8669-369c7c2258dc.png)

**Comment by wsalesky** [comment]
@mMoliere the latest changes to the main page are made in the version 1.04 version of the XSLT, make sure you are running that one. Also sometimes there is caching that happens with the forms, I always delete the old version of the form on my local version of eXist before uploading a new version, to make sure I get the latest of all the files.

**Comment by mMoliere** [comment]
@wsalesky The new version works like a charm!

Some more requests: 

- Captions of collapsibles should not disappear upon mouseover (rather change colour)
- Upon opening a subpage, all collapsibles should by default be closed/collapsed

**Comment by wsalesky** [comment]
@mMoliere See the latest XSLT for the updates. (version 1.05)

This addresses the changing text color on collapsibles and the having all collapsibles closed by default. 

I'm not sure what you mean by the last item. If you want the msItem in a separate collapsible, can you not add it in the config file?

**Comment by mMoliere** [comment]
@wsalesky Many thanks!

What I mean, all the `msItem` elements are displayed in one and the same collapsible. It would be much more easy for our editors to parse the page, if each `msItem` was in its separate collapsible.

**Comment by wsalesky** [comment]
I will see what I can do.

**Comment by wsalesky** [comment]
@mMoliere What is the logic behind this? As in, how do I know when an the whole collapsable section should be repeatable, rather than having the child elements within the collapsable being repeatable? Should we add a trigger to the config.xml? Or should I assume if there is only one child that the whole collapsable section should repeat? 

Thanks for the clarification.

**Comment by mMoliere** [comment]
@wsalesky I can provide a list of elements that we want to display as repeatable collapsibles:

Content: `msItem`
Text Layout: `decoNote`
Hand: `handNote`
history: `provenance`

**Comment by wsalesky** [comment]
@mMoliere I was hoping for some sort of logic that I could build on, not specific elements. If there is not I will add a trigger to the section in the config.xml file. Hope to have something to test by the end of this week.

**Comment by mMoliere** [comment]
@wsalesky I understand what you mean. Unfortunately, we cannot formulate a general rule.

**Comment by wsalesky** [comment]
@mMoliere I have added an attribute to the config.xml to indicate if a section is repeatable. Please test and let me know how it works.

**Comment by mMoliere** [comment]
@wsalesky That's great news! I shall test it and give you feedback

**Comment by wsalesky** [comment]
@mMoliere Here: https://github.com/majlis-erc/manuForma/blob/main/forms/formGenerator/config.xml#L192

**Comment by mMoliere** [comment]
@wsalesky Okay, I see what you did. That is a very convenient solution! I still can't see the effect in the browser. (I emptied my cache.)

**Comment by wsalesky** [comment]
@mMoliere Please see the latest version of the XSLT (1.06 Beta)

**Comment by wsalesky** [comment]
Fixed in version 1.07
